File: backend/core/src/modules/beta-sessions/beta-sessions.controller.ts
import { Body, Controller, Delete, Get, Param, Patch, Post, Query, Req, UseGuards } from '@nestjs/common';
import { JwtAuthGuard } from '../auth/jwt-auth.guard';
import { BetaSessionsService } from './beta-sessions.service';
import { CreateBetaSessionDto } from './dto/create-beta-session.dto';
import { UpdateBetaSessionDto } from './dto/update-beta-session.dto';
import { CreateBetaSessionInviteDto } from './dto/create-beta-session-invite.dto';
import { JoinBetaSessionDto } from './dto/join-beta-session.dto';

@UseGuards(JwtAuthGuard)
@Controller('beta-sessions')
export class BetaSessionsController {
  constructor(private readonly betaSessionsService: BetaSessionsService) {}

  @Get()
  list(@Req() req: any, @Query('projectId') projectId?: string) {
    return this.betaSessionsService.listForUser(req.user?.userId, projectId);
  }

  @Get('invite')
  inviteInfo(@Req() req: any, @Query('token') token: string) {
    return this.betaSessionsService.getInviteInfo(req.user?.userId, token);
  }

  @Post('join')
  join(@Req() req: any, @Body() dto: JoinBetaSessionDto) {
    return this.betaSessionsService.joinByInvite(req.user?.userId, dto.token);
  }

  @Get(':id')
  findOne(@Req() req: any, @Param('id') id: string) {
    return this.betaSessionsService.findOneForUser(req.user?.userId, id);
  }

  @Post()
  create(@Req() req: any, @Body() dto: CreateBetaSessionDto) {
    return this.betaSessionsService.createForUser(req.user?.userId, dto);
  }

  @Patch(':id')
  update(@Req() req: any, @Param('id') id: string, @Body() dto: UpdateBetaSessionDto) {
    return this.betaSessionsService.updateForUser(req.user?.userId, id, dto);
  }

  @Delete(':id')
  async remove(@Req() req: any, @Param('id') id: string) {
    await this.betaSessionsService.removeForUser(req.user?.userId, id);
    return { success: true };
  }

  @Post(':id/invites')
  createInvite(@Req() req: any, @Param('id') id: string, @Body() dto: CreateBetaSessionInviteDto) {
    return this.betaSessionsService.createInviteForSession(req.user?.userId, id, dto);
  }

  @Get(':id/participants')
  participants(@Req() req: any, @Param('id') id: string) {
    return this.betaSessionsService.listParticipantsForSession(req.user?.userId, id);
  }

  @Get(':id/document')
  document(@Req() req: any, @Param('id') id: string) {
    return this.betaSessionsService.getSessionDocumentForUser(req.user?.userId, id);
  }
}

File: backend/core/src/main/java/store/pjcloud/cowrite/core/controller/BetaSessionsController.java
package store.pjcloud.cowrite.core.controller;

import java.util.List;
import java.util.Map;
import java.util.UUID;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PatchMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;
import store.pjcloud.cowrite.core.entity.BetaSession;
import store.pjcloud.cowrite.core.entity.Document;
import store.pjcloud.cowrite.core.security.SecurityUtils;
import store.pjcloud.cowrite.core.service.BetaSessionsService;

@RestController
@RequestMapping("/beta-sessions")
public class BetaSessionsController {
    private final BetaSessionsService betaSessionsService;

    public BetaSessionsController(BetaSessionsService betaSessionsService) {
        this.betaSessionsService = betaSessionsService;
    }

    @GetMapping
    public List<BetaSession> list(@RequestParam(value = "projectId", required = false) UUID projectId) {
        UUID userId = SecurityUtils.requireUserId();
        return betaSessionsService.listForUser(userId, projectId);
    }

    @GetMapping("/invite")
    public Map<String, Object> inviteInfo(@RequestParam("token") String token) {
        UUID userId = SecurityUtils.requireUserId();
        return betaSessionsService.getInviteInfo(userId, token);
    }

    @PostMapping("/join")
    public Map<String, Object> join(@RequestBody Map<String, Object> body) {
        UUID userId = SecurityUtils.requireUserId();
        String token = body.get("token") instanceof String s ? s : null;
        return betaSessionsService.joinByInvite(userId, token);
    }

    @GetMapping("/{id}")
    public BetaSession findOne(@PathVariable("id") UUID id) {
        UUID userId = SecurityUtils.requireUserId();
        return betaSessionsService.findOneForUser(userId, id);
    }

    @PostMapping
    public BetaSession create(@RequestBody Map<String, Object> body) {
        UUID userId = SecurityUtils.requireUserId();
        return betaSessionsService.createForUser(userId, body);
    }

    @PatchMapping("/{id}")
    public BetaSession update(@PathVariable("id") UUID id, @RequestBody Map<String, Object> body) {
        UUID userId = SecurityUtils.requireUserId();
        return betaSessionsService.updateForUser(userId, id, body);
    }

    @DeleteMapping("/{id}")
    public Map<String, Object> remove(@PathVariable("id") UUID id) {
        UUID userId = SecurityUtils.requireUserId();
        betaSessionsService.removeForUser(userId, id);
        return Map.of("success", true);
    }

    @PostMapping("/{id}/invites")
    public Map<String, Object> createInvite(@PathVariable("id") UUID id, @RequestBody Map<String, Object> body) {
        UUID userId = SecurityUtils.requireUserId();
        return betaSessionsService.createInviteForSession(userId, id, body);
    }

    @GetMapping("/{id}/participants")
    public List<Map<String, Object>> participants(@PathVariable("id") UUID id) {
        UUID userId = SecurityUtils.requireUserId();
        return betaSessionsService.listParticipantsForSession(userId, id);
    }

    @GetMapping("/{id}/document")
    public Document document(@PathVariable("id") UUID id) {
        UUID userId = SecurityUtils.requireUserId();
        return betaSessionsService.getSessionDocumentForUser(userId, id);
    }
}
